File: impl/app/android/app/src/main/java/com/ak/contexta/domain/error/DomainResult.kt
package com.ak.contexta.domain.error

/**
 * 同步操作的统一返回类型，用于 Use Case 和 Repository 方法。
 * ViewModel 根据结果转为 UI 状态（Snackbar、Dialog 等）。
 */
sealed class DomainResult<out T> {
    data class Success<T>(val data: T) : DomainResult<T>()
    data class Failure(val error: AppError) : DomainResult<Nothing>()

    val isSuccess: Boolean get() = this is Success
    val isFailure: Boolean get() = this is Failure

    fun getOrNull(): T? = when (this) {
        is Success -> data
        is Failure -> null
    }

    fun errorOrNull(): AppError? = when (this) {
        is Success -> null
        is Failure -> error
    }

    inline fun onSuccess(action: (T) -> Unit): DomainResult<T> {
        if (this is Success) action(data)
        return this
    }

    inline fun onFailure(action: (AppError) -> Unit): DomainResult<T> {
        if (this is Failure) action(error)
        return this
    }
}

File: impl/app/flutter/lib/domain/error/domain_result.dart
import 'app_error.dart';

/// 同步操作的统一返回类型（对齐 Kotlin DomainResult.kt），
/// 用于 Use Case 和 Repository 方法。UI 根据结果转为界面状态
/// （Snackbar、Dialog 等）。
///
/// Kotlin 中为 sealed class（非 stdlib Result），Dart 对应 sealed class。
sealed class DomainResult<T> {
  const DomainResult();

  bool get isSuccess => switch (this) {
        Success<T>() => true,
        Failure() => false,
      };

  bool get isFailure => !isSuccess;

  /// 成功时返回数据，失败返回 null（对齐 Kotlin `getOrNull()`）。
  T? get dataOrNull => switch (this) {
        Success<T>(:final data) => data,
        Failure() => null,
      };

  /// 失败时返回错误，成功返回 null（对齐 Kotlin `errorOrNull()`）。
  AppError? get errorOrNull => switch (this) {
        Success<T>() => null,
        Failure(:final error) => error,
      };

  /// 成功时执行 [action]，返回自身便于链式调用。
  DomainResult<T> onSuccess(void Function(T data) action) {
    if (this case Success<T>(:final data)) action(data);
    return this;
  }

  /// 失败时执行 [action]，返回自身便于链式调用。
  DomainResult<T> onFailure(void Function(AppError error) action) {
    if (this case Failure(:final error)) action(error);
    return this;
  }
}

/// 成功分支，携带泛型数据。
class Success<T> extends DomainResult<T> {
  final T data;

  const Success(this.data);
}

/// 失败分支，携带 [AppError]（对齐 Kotlin `DomainResult<Nothing>`）。
class Failure extends DomainResult<Never> {
  final AppError error;

  const Failure(this.error);
}
